<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-цели-и-задач" data-toc-modified-id="Описание-цели-и-задач-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание цели и задач</a></span></li><li><span><a href="#Импорт-библиотек-и-подключение-к-базе-данных" data-toc-modified-id="Импорт-библиотек-и-подключение-к-базе-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт библиотек и подключение к базе данных</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Задачи" data-toc-modified-id="Задачи-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задачи</a></span><ul class="toc-item"><li><span><a href="#Задание-1" data-toc-modified-id="Задание-1-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Задание 1</a></span></li><li><span><a href="#Задание-2" data-toc-modified-id="Задание-2-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Задание 2</a></span></li><li><span><a href="#Задание-3" data-toc-modified-id="Задание-3-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Задание 3</a></span></li><li><span><a href="#Задание-4" data-toc-modified-id="Задание-4-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Задание 4</a></span></li><li><span><a href="#Задание-5" data-toc-modified-id="Задание-5-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Задание 5</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Проект по SQL

### Описание цели и задач

В связи с популярностью книг на фоне коронавируса и распространенной самоизоляции, чтобы сформулировать ценностное предложение для нового продукта для развития крупного сервиса для чтения книг по подписке, необходимо проанализировать базу данных, в которой содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

**Цель исследования** — проанализировать базу данных для формирования ценностных предложений для нового продукта.

**Описание данных**

Таблица `books`
Содержит данные о книгах:
* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

Таблица `authors`
Содержит данные об авторах:
* `author_id` — идентификатор автора;
* `author` — имя автора.

Таблица `publishers`
Содержит данные об издательствах:
* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства.

Таблица `ratings`
Содержит данные о пользовательских оценках книг:
* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

Таблица `reviews`
Содержит данные о пользовательских обзорах:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя автора обзора;
* `text` — текст обзора.

**Схема базы данных**

![Схема базы данных][1]

[1]: https://pictures.s3.yandex.net/resources/scheme_1589269096.png "Схема базы данных"

### Импорт библиотек и подключение к базе данных

Импортируем библиотеки

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

Устанавливаем параметры соединения (`db_config`):
* `user` - имя пользователя, 
* `pwd` - пароль, 
* `host` - хост, 
* `port` - порт, 
* `db` - название базы данных

In [3]:
# создаем переменную строка подключения
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Обзор данных

Выведем первые пять строк каждой из таблиц базы данных

In [5]:
# выводим первые пять строк books
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
# выводим первые пять строк authors
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
# выводим первые пять строк ratings
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
# выводим первые пять строк reviews
query = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [9]:
# выводим первые пять строк publishers
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Задачи

#### Задание 1

**Посчитайте, сколько книг вышло после 1 января 2000 года**

In [10]:
# решение задачи 1
query = '''SELECT COUNT(book_id) FROM books WHERE CAST(publication_date AS date) > '2000-01-01';'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

В базе содержится информация о **819** книгах, которые вышли после 1 января 2000 года.

#### Задание 2

**Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [11]:
# решение задачи 2
query = '''SELECT b.book_id, COUNT(DISTINCT re.review_id), AVG(ra.rating)
FROM books AS b 
LEFT JOIN ratings AS ra ON b.book_id=ra.book_id
LEFT JOIN reviews AS re ON b.book_id=re.book_id 
GROUP BY b.book_id'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  count       avg
0          1      2  3.666667
1          2      1  2.500000
2          3      3  4.666667
3          4      2  4.500000
4          5      4  4.000000
..       ...    ...       ...
995      996      3  3.666667
996      997      3  3.400000
997      998      4  3.200000
998      999      2  4.500000
999     1000      4  3.833333

[1000 rows x 3 columns]

Для каждой книги посчитаны: количество обзоров и средняя оценка.

#### Задание 3

**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [12]:
# решение задачи 3
query = '''SELECT p.publisher, COUNT(b.book_id)
FROM publishers AS p
LEFT JOIN books AS b ON p.publisher_id=b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY count DESC
LIMIT 1;'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  count
0  Penguin Books     42

Издательство, которое выпустило больше всего книг толще 50 страниц (**42 книги**) - **Penguin Books**

#### Задание 4

**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [13]:
# решение задачи 4
query = '''SELECT a.author, AVG(ra.rating)
FROM (SELECT book_id
      FROM ratings
      GROUP BY book_id
      HAVING COUNT(rating_id) > 50) AS ch
LEFT JOIN books AS b ON ch.book_id=b.book_id
LEFT JOIN authors AS a ON b.author_id=a.author_id
LEFT JOIN ratings AS ra ON b.book_id=ra.book_id
GROUP BY a.author_id
ORDER BY avg DESC
LIMIT 1'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author       avg
0  J.K. Rowling/Mary GrandPré  4.287097

Автор с самой высокой средней оценкой книг (учтены только книги с 50 и более оценками) - **J.K. Rowling/Mary GrandPré**, а рейтинг составляет **~4.3**

#### Задание 5

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок**

In [14]:
# решение задачи 5
query = '''SELECT AVG(rew.count_rew)
FROM (SELECT username
      FROM ratings
      GROUP BY username
      HAVING COUNT(rating_id) > 48) AS u
LEFT JOIN (SELECT username, COUNT(review_id) AS count_rew
           FROM reviews
           GROUP BY username) AS rew
ON u.username=rew.username'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок составляет **24**

### Выводы

С помощью SQL-запросов была проанализирована база данных для формирования ценностных предложений для нового продукта.
Получено:
* В базе содержится информация о **819** книгах, которые вышли после 1 января 2000 года.
* Для каждой книги посчитаны: количество обзоров и средняя оценка
* Издательство, которое выпустило больше всего книг толще 50 страниц (**42 книги**) - **Penguin Books**
* Автор с самой высокой средней оценкой книг (учтены только книги с 50 и более оценками) - **J.K. Rowling/Mary GrandPré**, а рейтинг составляет **~4.3**
* Среднее количество обзоров от пользователей, которые поставили больше 48 оценок составляет **24**